### ISSUE TITLE
Revisit EINTR handling with check_interrupts = TRUE

### ISSUE THREAD

**Issue by krlmlr** [r-dbi/RPostgres#344]
> * [Linux man on select](https://man7.org/linux/man-pages/man2/select.2.html#NOTES) states that the `EINTR` should be interpreted as "your select() call has been interrupted, try again". Which is in contrary to your current implementation.

From https://github.com/r-dbi/RPostgres/pull/337#issuecomment-922844813.

@zozlak: I don't follow where the man page you pointed to suggests that it's mandatory to call `select()` again. Can you please clarify?

**Comment by zozlak** [comment]
Last sentence o the third paragraph of the notes section states (citing whole paragraph):

> On some other UNIX systems, select() can fail with the error EAGAIN if the system fails to allocate kernel-internal resources, rather than ENOMEM as Linux does.  POSIX specifies this error for poll(2), but not for select().  Portable programs may wish to check for EAGAIN and loop, just as with EINTR.

**Comment by zozlak** [comment]
The thing is (in contrary to the Ms Windows documentation) there is no list of possible signals which can interrupt the `select()` and the [list of standard signals](https://man7.org/linux/man-pages/man7/signal.7.html) (section "Standard signals") is long and includes a lot of stuff which we just don't care about while waiting for the response from the Postgresql server. Let's say someone sent a `USR1` signal and it caused `select()` to return with the `EINTR` (I don't know if it's possible in practice and I'm too lazy to go into the kernel code of all *nixes R can run on to check it but the documentation I'm referring to warns us that in general everything is possible and a portable code should take it into account). Should we abandon SQL query execution because of that? I would say now and we should just repeat the `select()`.

**Comment by zozlak** [comment]
First, we only care about whatever R considers a user request for execution interruption. This may or may not correspond to the `SIGINT` signal. IMHO we don't want to think about it. We definitely want some abstraction level here and as far as I can tell that's what `Rcpp::checkUserInterrupt()` provides.

Second, the documentation says only that the `select()` may be interrupted by any signal. It doesn't mean it does is interrupted by any particular signal on any particular platform and the lesson learned from the problems we encountered with completely different `select()` behavior on Windows is that the "if it works like that on my machine, it must work like that everywhere" approach is a bad approach :-) . Therefore we shouldn't make the assumption that a particular signal like `SIGINT` interrupts the `select()`. Anyway, IMHO we shouldn't think about it in terms of OS signals in the first place (see the previous paragraph).

Summing up on *nix systems I would treat `select()` returning the `EINTR` in the same way as `select()` returning 0 - just call the `try{Rcpp::checkUserInterrupt();}catch(...){return false;}` and call the `select()` again. If the `select()` has been interrupted by an event which will be recognized as the request to terminate the execution by the `Rcpp::checkUserInterrupt()`, it will terminate and if it was any other event we don't care about here, we will continue waiting.

**Comment by krlmlr** [comment]
Thanks for your help!

**Comment by krlmlr** [comment]
Need to test this in 20.04. If the 21.10 behavior is different, we might need to take action.

**Comment by krlmlr** [comment]
Also, good catch regarding the mismatch between test description and feature tested.

**Comment by krlmlr** [comment]
Are you testing this on RStudio Server? What is the `rsession` process?

In RStudio, pressing Escape in the console sends an interrupt.

I installed Ubuntu 22.04, it is showing the same behavior. I tweaked RPostgres to show a dedicated error message (`Interrupted.`) if an interrupt is detected. The tests pass (except on Windows, https://github.com/r-lib/processx/issues/319)  -- they currently use `session$interrupt()` and I also tried `ps::ps_send_signal(ps::ps_handle(session$get_pid()), 2L)` with the same result.

In my tests I haven't seen `Rcpp::checkUserInterrupt()` being used at all.

I will fix the description of the tests, take another stab at stabilizing the tests on Windows (following Gábor's input) and then close this.

**Comment by zozlak** [comment]
> Are you testing this on RStudio Server? What is the rsession process?

Local RStudio but I tried also with pure R (adjusting the `kil`l call accordingly) and the outcome was exactly the same.

> In RStudio, pressing Escape in the console sends an interrupt.

I preferred `kill` so I know exactly which interrupt I send. Anyway I can replicate the same behavior (lack of both `SOCKET_EINTR` and `checkUserInterrupt after timeout` messages) also while using `Escape` in RStudio and `CTRL+C` in R run in shell.

It looks like while `select()` is waiting for timeout/data R's event loop responsible for handling user interrupts can do its job. On one hand it's good - we don't need to care about checking user interrupts on our own - on the other it's bad - we can't assure `pConnPtr_->cancel_query();` is called upon user's interrupt.

> The tests pass (...)

Yes, it passes the test but the `pConnPtr_->cancel_query();` isn't called (I checked it). It means while it seems to work on the R side (`dbGetQuery()` ends - everything must be fine) it leaves the query running on the Postgresql server. As the main use case for breaking query execution is "I run too complex query which will not execute in the sane time", leaving such queries running on the server is definitely a big shortcoming. The really good test should use a side Postgresql connection reading from the `pg_stat_activity` table and making sure the query has been cancelled. I will try to prepare it in a moment.

**Comment by zozlak** [comment]
A better test - https://github.com/zozlak/RPostgres/blob/issue-344/tests/testthat/test-checkInterrupts.R#L9-L63

**Comment by zozlak** [comment]
The solution can be something like (sorry for a quick & dirty sudo-code):

```R
DbGetQuery = function() {
  result_handle = initialize_query_object()
  result_handle$should_be_cancelled = TRUE
  on.exit({
    if (result_handle$should_be_cancelled) {
      call_cancel_query_on_result_handle()
    }
  })
  run_the_query()
  wait_for_data()
  results = fetch_data()
  result_handle$should_be_cancelled = FALSE
  results
}
```
because code registered with `on.exit()` is still run by R when an R function is interrupted by the user interrupt.

**Comment by krlmlr** [comment]
Good point, let's add checks to verify if `cancel_query()` is called.

**Comment by krlmlr** [comment]
I ran the following script on my system to confirm that `cancel_query()` is called:

```r
pkgload::load_all()

init_logging("DEBUG")

.GlobalEnv$conn <- postgresDefault(check_interrupts = TRUE)

dbGetQuery(.GlobalEnv$conn, "SELECT pg_sleep(10)")
```

During the sleep, I pressed Esc in RStudio and Ctrl+C in the shell. In both cases the following output appeared:

```
> source("~/git/R/r-dbi/RPostgres/intr.R", echo=TRUE)

> pkgload::load_all()
ℹ Loading RPostgres

> init_logging("DEBUG")

> .GlobalEnv$conn <- postgresDefault(check_interrupts = TRUE)
PqResultImpl::PqResultImpl@34: SET datestyle to 'iso, mdy'
void PqResultImpl::bind@275: 0
bool PqResultImpl::wait_for_data@586: 1
bool PqResultImpl::wait_for_data@586: 1
PqResultImpl::PqResultImpl@34: SET TIMEZONE = 'UTC'
void PqResultImpl::bind@275: 0
bool PqResultImpl::wait_for_data@586: 1
bool PqResultImpl::wait_for_data@611: 
bool PqResultImpl::wait_for_data@586: 1
PqResultImpl::PqResultImpl@34: SELECT oid, typname FROM pg_type
void PqResultImpl::bind@275: 0
bool PqResultImpl::wait_for_data@586: 1
bool PqResultImpl::wait_for_data@611: 
void PqResultImpl::_cache::set@95: 0
Rcpp::List PqResultImpl::fetch@308: -1
Rcpp::List PqResultImpl::fetch_rows@424: -1/0
bool PqResultImpl::wait_for_data@586: 1

> dbGetQuery(.GlobalEnv$conn, "SELECT pg_sleep(10)")
PqResultImpl::PqResultImpl@34: SELECT pg_sleep(10)
void PqResultImpl::prepare@229: SELECT pg_sleep(10)
void PqResultImpl::_cache::set@95: 0
void PqResultImpl::bind@275: 0
bool PqResultImpl::wait_for_data@586: 1
bool PqResultImpl::wait_for_data@611: 
bool PqResultImpl::wait_for_data@621: 
bool PqResultImpl::wait_for_data@611: 
bool PqResultImpl::wait_for_data@635: 
void DbConnection::cancel_query@92: 
void DbConnection::cancel_query@103: 
Error: Interrupted.
```

This confirms that `PQcancel()` is called.

Define `R_PLOGR_ENABLE=1` in your `.Renviron` before running the script to enable logging.

Please open a new issue if you find that `cancel_query()` isn't called in some circumstances.

Thanks for pointing me to improved test versions. Checking query state on the server feels brittle, can we enrich the query with a nonce (e.g. a GUID that is part of the query string) to ensure that the results are unique? Unfortunately, the patch no longer applies.

**Comment by zozlak** [comment]
> Checking query state on the server feels brittle, can we enrich the query with a nonce (e.g. a GUID that is part of the query string) to ensure that the results are unique?

Sure, we can get the pid of the process handling a given connection with `SELECT pg_backend_pid();`.
I will prepare a nicer test later on today.

> Define R_PLOGR_ENABLE=1 in your .Renviron before running the script to enable logging.

Many thanks for that hint! I noticed there is a nicer debugging built into the code but was too lazy to check how to enable it.
